### PR TITLE
CMP: fix flag indicating the use of SSL/TLS for client-side HTTP connections

### LINKED ISSUE #21120: Why is http_cb_arg used to determine whether to use tls in the CMP?
OpenSSL version：3.0
In OSSL_CMP_MSG_http_perform(),  use http_cb_arg to determine whether to use TLS. If http_cb is set but http_cb_arg is not set, the selection of the default port and https_proxy will be affected.
Must both http_cb and http_cb_arg be set?

**@crypto/cmp/cmp_http.c 71**
`tls_used = OSSL_CMP_CTX_get_http_cb_arg(ctx) != NULL;`

### PR DESCRIPTION
**Pull request by DDvO**
* CMP: fix `OSSL_CMP_MSG_http_perform()` by adding option `OSSL_CMP_OPT_USE_TLS`
* `OSSL_CMP_CTX_new.pod`: remove overlap with `OSSL_HTTP_transfer.pod`; improve the latter
* `apps/cmp.c`: `-tls_used` may be implied by `-server https:...`; improve related checks and doc

Fixes #21120 

##### Checklist

- [x] documentation is  added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
The two CI failures are unrelated.

**Comment by t8m** [comment]
> The two CI failures are unrelated.

Although unrelated the CMS failure looks really suspicious.

**Comment by DDvO** [comment]
> > The two CI failures are unrelated.
> 
> Although unrelated the CMS failure looks really suspicious.

I agree the failing _CMS consistency tests_ on **buildbot/master:unix-ubuntu-aarch64** looked suspicious,
but luckily this disappeared after I rebased to the latest `master` half an hour ago.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man1/openssl-cmp.pod.in [hunk 1]
```diff
@@ -841,14 +841,14 @@ B<-tls_key>.
...
 =item B<-tls_used>
 
-Enable using TLS (even when other TLS-related options are not set)
-for message exchange with CMP server via HTTP.
+Make the CMP client use TLS (regardless if other TLS-related options are set)
+for message exchange with the server via HTTP.
 This option is not supported with the I<-port> option.
-It is ignored if the I<-server> option is not given or I<-use_mock_srv> is given
-or I<-rspin> is given with enough filename arguments.
+It is implied of the B<-server> option is given with the scheme C<https>.
```
[the remark this answers is not in the document]
Comment (DDvO) [reply doc/man1/openssl-cmp.pod.in]: Oops, fixed.

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
I do not think this is a bug fix. Also what if anyone is already depending on the existing behavior of having the callback implying TLS?

File: crypto/cmp/cmp_local.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ struct ossl_cmp_ctx_st {
     int keep_alive; /* persistent connection: 0=no, 1=prefer, 2=require */
     int msg_timeout; /* max seconds to wait for each CMP message round trip */
     int total_timeout; /* max number of seconds an enrollment may take, incl. */
+    int tls_used; /* whether to use TLS for client-side HTTP connections */
```
Comment (tmshort) [inline crypto/cmp/cmp_local.h:52]: Presumably, this is not just for clients? In `apps/cmp.c` the flag is set regardless of `opt_server`

**Comment by DDvO** [comment]
> I do not think this is a bug fix.

As witnessed by #21120, the existing code has at least unexpected and undocumented behavior.
I see it as a bug because the use of the callback and its optional argument is needlessly limited to adding TLS (to the given CMP HTTP transfer).
If it is used in the CMP context for anything else, as can be done for the general (non-CMP) HTTP client, things go wrong as it is falsely assumed that TLS is being used (if the argument is non-NULL) or not being used (if the arg is NULL).

> Also what if anyone is already depending on the existing behavior of having the callback implying TLS?

This is indeed a problem, but I'd say the risk is small because it's an advanced feature and the existing behavior is weird and not documented.

**Review by tmshort — CHANGES_REQUESTED** [review]
Just a clarification needed on server vs client use of `OSSL_CMP_OPT_USE_TLS`.

File: crypto/cmp/cmp_local.h (hunk 2 above)
Comment (DDvO) [reply crypto/cmp/cmp_local.h:52]: So far, `tls_used` is implemented for the client side only.
And as now stated in  `openssl-cmp.h.in`:
```
It is ignored if the B<-server> option is not given
```

**Comment by t8m** [comment]
> > Also what if anyone is already depending on the existing behavior of having the callback implying TLS?
> 
> With this argument, we couldn't do any bug fixes because there might be someone who - for whatever strange reason - started relying on wrong behavior.

The app code already did this, what if someone just mimicked it? Can you please try harder to make this backwards compatible?

For example this could be a new feature for 3.2 where the OP is OSSL_CMP_OPT_NO_TLS and it switches off the tls if the callback is in use.

**Comment by DDvO** [comment]
> > > Also what if anyone is already depending on the existing behavior of having the callback implying TLS?
> > 
> > With this argument, we couldn't do any bug fixes because there might be someone who - for whatever strange reason - started relying on wrong behavior.
> 
> The app code already did this, what if someone just mimicked it? Can you please try harder to make this backwards compatible?
> 
> For example this could be a new feature for 3.2 where the OP is OSSL_CMP_OPT_NO_TLS and it switches off the tls if the callback is in use.

Hmm, such a `OSSL_CMP_OPT_NO_TLS` option would only solve half of the problem.

Yet here is backward compatible solution:
define for the new `OSSL_CMP_OPT_USE_TLS` option a third value, namely -1, which is assumed by the default and means the hitherto behavior, where the callback function arg determines whether to use TLS.
When the `OSSL_CMP_OPT_USE_TLS` option is set to 0, TLS is not used, and when set to 1, TLS is used.

Ok?

**Comment by t8m** [comment]
OK, but for master branch only.

**Comment by DDvO** [comment]
Thanks @mattcaswell, @tmshort, and @t8m for your comments.
I've now implemented the newly proposed backward compatible solution for fixing #21120.

**Review by tmshort — APPROVED** [review]
I'm ok with a back port to 3.1/3.0. Should this also be in master (not labeled for master, but the PR is on master).

**Comment by DDvO** [comment]
> I'm ok with a back port to 3.1/3.0. Should this also be in master (not labeled for master, but the PR is on master)?

Pleased to hear.
I simply forgot to tick the `master` label - done now.

**Comment by DDvO** [comment]
Ping @openssl/committers for 2nd review

**Comment by DDvO** [comment]
Pinging again @openssl/committers for 2nd review.

**Review by t8m — APPROVED** [review]
OK for the master branch only.

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**